scg-system/src/main/java/io/telicent/backup/services/DatasetBackupService.java

```diff
@@ -0,0 +1,276 @@
+        ObjectNode response = MAPPER.createObjectNode();
+        response.put("date", DateTimeUtils.nowAsString("yyyy-MM-dd_HH-mm-ss"));
+        response.set("backups", populateNodeFromDirNumerically(getBackUpDir()));
+        return response;
+    }
+
+    public ObjectNode restoreDatasets(String restoreId) {
+        ObjectNode response = MAPPER.createObjectNode();
+        String restorePath = getBackUpDir() + "/" + restoreId;
+        response.put("restorePath", restorePath);
+        if (!checkPathExistsAndIsDir(restorePath)) {
+            response.put("reason", "Restore path unsuitable: " + restorePath);
+            response.put("success", false);
+        } else {
+            List<String> datasets = getSubdirectoryNames(restorePath);
+            if (datasets.isEmpty()) {
+                response.put("reason", "Restore path unsuitable: " + restorePath);
+                response.put("success", false);
+            } else {
+                for (String datasetName : datasets) {
+                    response.set(datasetName, restoreDataset(restorePath, datasetName));
+                }
+            }
+        }
+        return response;
+    }
+
+    ObjectNode restoreDataset(String restorePath, String datasetName) {
+        ObjectNode response = MAPPER.createObjectNode();
+        response.put("dataset-id", datasetName);
+        DataAccessPoint dataAccessPoint = dapRegistry.get("/" + datasetName);
+        if (dataAccessPoint == null) {
+            response.put("reason", datasetName + " does not exist");
+            response.put("success", false);
+            return response;
+        }
+        DatasetGraph dsg = dataAccessPoint.getDataService().getDataset();
+        response.set("tdb", restoreTDB(datasetName, dsg, restorePath));
+        response.set("labels", restoreLabelStore(datasetName, dsg, restorePath));
+        return response;
+    }
+
+    ObjectNode restoreTDB(String datasetName, DatasetGraph dsg, String restorePath) {
+        ObjectNode node = MAPPER.createObjectNode();
+        String tdbRestoreFile = restorePath + "/" + datasetName + "/TDB/" + datasetName + "_backup.nq.gz";
+        node.put("restorePath", tdbRestoreFile);
+        if (!checkPathExistsAndIsFile(tdbRestoreFile)) {
+            node.put("reason", "Restore file not found: " + tdbRestoreFile);
+            node.put("success", false);
+            return node;
+        }
+        try {
+            executeRestoreTDB(dsg, tdbRestoreFile);
+            node.put("success", true);
+        } catch (Exception e) {
+            node.put("reason", e.getMessage());
+            node.put("success", false);
+        }
+        return node;
+    }
+
+    @ExcludeFromJacocoGeneratedReport
+    void executeRestoreTDB(DatasetGraph dsg, String tdbRestoreFile) throws IOException {
+        try (InputStream fis = new FileInputStream(tdbRestoreFile);
+             InputStream gis = new GZIPInputStream(fis)) {
+            Txn.executeWrite(dsg, () -> {
```

Comment:
For future ticket, we may want to put the `Txn.executeWrite()` around the entire restoration (data + labels) in the `restoreDataset()` method as otherwise if one restore succeeds and the other fails you'd have inconsistent database states.

Reply:
Good idea.